### PR TITLE
Alerting: Support for optimistic concurrency in priovisioning Tempate API

### PR DESCRIPTION
**Pull request by yuri-tceretian**
**What is this feature?**
This PR updates the template provisioning API to support optional optimistic concurrency. 
 - PUT method updated to accept optional field `version` in the payload. If the version is not empty and the template does not exist, then API will return `404 NotFound` 
 - DELETE method is updated to accept optional query parameters`version`.

Documentation has been updated to mention changes in template and mute timing APIs (https://github.com/grafana/grafana/pull/88731)

**Why do we need this feature?**
Support optimistic concurrency for each resource and prepare for migration to K8s API.

**Who is this feature for?**
Provisioning API users and UI users in the future

Please check that:
- [x] It works as expected from a user's perspective.
- [x] If this is a pre-GA feature, it is behind a feature toggle.
- [x] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

File: docs/sources/shared/alerts/alerting_provisioning.md [hunk 1]
```diff
@@ -1568,9 +1608,10 @@ Status: Accepted
 
 {{% responsive-table %}}
 
-| Name     | Type   | Go type  | Required | Default | Description | Example |
-| -------- | ------ | -------- | :------: | ------- | ----------- | ------- |
-| template | string | `string` |          |         |             |         |
+| Name     | Type   | Go type  | Required | Default | Description                                            | Example |
+| -------- | ------ | -------- | :------: | ------- | ------------------------------------------------------ | ------- |
+| template | string | `string` |          |         |                                                        |         |
+| version  | string | `string` |          |         | version of resource. Should be empty for new templates |         |
```
Comment (brendamuir) [inline docs/sources/shared/alerts/alerting_provisioning.md:1614]:
```suggestion
| version  | string | `string` |          |         | Version of resource. Should be empty for new templates. |         |
```

File: docs/sources/shared/alerts/alerting_provisioning.md [hunk 2]
```diff
@@ -1516,10 +1554,11 @@ Status: Accepted
...
 
-| Name           | Type                             | Go type           | Required | Default | Description | Example |
-| -------------- | -------------------------------- | ----------------- | :------: | ------- | ----------- | ------- |
-| name           | string                           | `string`          |          |         |             |         |
-| time_intervals | [][TimeInterval](#time-interval) | `[]*TimeInterval` |          |         |             |         |
+| Name           | Type                             | Go type           | Required | Default | Description         | Example |
+| -------------- | -------------------------------- | ----------------- | :------: | ------- | ------------------- | ------- |
+| name           | string                           | `string`          |          |         |                     |         |
+| time_intervals | [][TimeInterval](#time-interval) | `[]*TimeInterval` |          |         |                     |         |
+| version        | string                           | `string`          |          |         | version of resource |         |
```
Comment (brendamuir) [inline docs/sources/shared/alerts/alerting_provisioning.md:1561]:
```suggestion
| version        | string                           | `string`          |          |         | Version of resource |         |
```

File: docs/sources/shared/alerts/alerting_provisioning.md [hunk 3]
```diff
@@ -1554,11 +1593,12 @@ Status: Accepted
...
-| ---------- | ------------------------- | ------------ | :------: | ------- | ----------- | ------- |
-| name       | string                    | `string`     |          |         |             |         |
-| provenance | [Provenance](#provenance) | `Provenance` |          |         |             |         |
-| template   | string                    | `string`     |          |         |             |         |
+| Name       | Type                      | Go type      | Required | Default | Description         | Example |
+| ---------- | ------------------------- | ------------ | :------: | ------- | ------------------- | ------- |
+| name       | string                    | `string`     |          |         |                     |         |
+| provenance | [Provenance](#provenance) | `Provenance` |          |         |                     |         |
+| template   | string                    | `string`     |          |         |                     |         |
+| version    | string                    | `string`     |          |         | version of resource |         |
```
Comment (brendamuir) [inline docs/sources/shared/alerts/alerting_provisioning.md:1601]:
```suggestion
| version    | string                    | `string`     |          |         | Version of resource |         |
```

File: docs/sources/shared/alerts/alerting_provisioning.md [hunk 4]
```diff
@@ -1758,3 +1799,18 @@ Status: Accepted
...
+**Properties**
+
+{{% responsive-table %}}
+
+| Name       | Type       | Go type          | Required | Default | Description                                                     | Example |
+| ---------- | ---------- | ---------------- | :------: | ------- | --------------------------------------------------------------- | ------- |
+| statusCode | string     | `string`         |    ✓     |         | HTTP Status Code                                                |         |
+| messageId  | string     | `string`         |    ✓     |         | Unique code of the error                                        |         |
+| message    | string     | `string`         |          |         | Error message                                                   |         |
+| extra      | map of any | `map[string]any` |          |         | Extra information about error. Format is specific to error code |         |
```
Comment (brendamuir) [inline docs/sources/shared/alerts/alerting_provisioning.md:1814]:
```suggestion
| extra      | map of any | `map[string]any` |          |         | Extra information about the error. Format is specific to the error code. |         |
```

**Review by brendamuir — APPROVED** [review]

File: docs/sources/shared/alerts/alerting_provisioning.md [hunk 5]
```diff
@@ -457,15 +457,17 @@ DELETE /api/v1/provisioning/mute-timings/:name
...
+| name    | `path`  | string | `string` |           |    ✓     |         | Mute timing name                                                                                              |
+| version | `query` | string | `string` |           |          |         | Current version of the resource. Used for optimistic concurrency validation. Keep empty to bypass validation. |
 
 #### All responses
 
 | Code                                 | Status     | Description                               | Has headers | Schema                                         |
 | ------------------------------------ | ---------- | ----------------------------------------- | :---------: | ---------------------------------------------- |
 | [204](#route-delete-mute-timing-204) | No Content | The mute timing was deleted successfully. |             | [schema](#route-delete-mute-timing-204-schema) |
+| [409](#route-delete-mute-timing-400) | Conflict   | GenericPublicError                        |             | [schema](#route-delete-mute-timing-409-schema) |
```
Comment (alexander-akhmetov) [inline docs/sources/shared/alerts/alerting_provisioning.md:470]: A typo here and later in the doc? `route-delete-mute-timing-409`

**Review by alexander-akhmetov — APPROVED** [review]
LGTM

**Comment by diegommm** [comment]
Hi! Greetings from the Release Guild. This is a friendly reminder that the Steady [11.2.x Release Branch](https://github.com/grafana/grafana/tree/v11.2.x) was cut this week, and that it includes the changes committed up to https://github.com/grafana/grafana/commit/b4126d3bce3627e9460e2db0e870c8d984be8e84. This PR was merged after it, so we wanted to let you know that you can add the `backport v11.2.x` label to to backport it and make it part of the next 11.2.x Patch Release, just in case you may have missed the news.

Thank you!

**Comment by yuri-tceretian** [comment]
Greetings from Release Guild. This PR has been successfully included in the Grafana 11.2 release branch. A backport PR was also created to ensure the v11.2.x branch is correctly synchronized with the release.

**Pull request merged**